#163298000 Validate username input

d-rita opened the pull request:
**What does this PR do?**
Implement validation errors for failed registration by a new user.

**Description of Task to be completed?**

- Validate the username input by new user
- Write descriptive error messages for this validation, i.e. "Username can only contain letters and numbers."
- Write a test to test the functionality of this validation. 

**How should this be manually tested?**
- Fetch this branch using `git fetch origin bg-fix-validate-username-163298000`
- Check out the branch using `git checkout bg-fix-validate-username-163298000`
- Create virtual environment using `virtualenv <envname>` and activate it using `source <envname>/bin/activate`
- To start the application, run `python manage.py runserver` in terminal.
- Open postman, and go to `http://localhost:8000/api/users/`. Use the POST method.
- Enter raw JSON data in the format:
`{ "user":{ "email": "user@myemail.com", "username":"myname", "password":"Secr3tc0de" } }`
- Enter data for an invalid username, i.e. a username containing spaces and special characters.

**Screenshot**
- Special Character in the username
<img width="1078" alt="special character" src="https://user-images.githubusercontent.com/31903212/51338706-9bf09880-1a9b-11e9-9722-d0c39477f5fc.png">
- Spaces in the username
<img width="1074" alt="screen shot 2019-01-17 at 9 05 39 pm" src="https://user-images.githubusercontent.com/31903212/51338741-b7f43a00-1a9b-11e9-8bcc-5419e66e28fe.png">

**Related Pivotal Tracker Stories**
[#163298000](#163298000)
[#162725254](#162725254)

ja-odur commented:
@d-rita Please update the link to PT story to point to the exact story.

ja-odur commented on authors/apps/authentication/serializers.py:
```diff
@@ -4,7 +4,7 @@
 import re
 from .models import User
-from authors.apps.utils.validators.validation_helpers import validate_password
+from authors.apps.utils.validators.validation_helpers import validate_password, valid_username
```
@d-rita there is a typo here. I think you meant `validate_username` instead of `valid_username`

d-rita replied:
Nope. the name of my function is valid_username as imported from the validation_helpers file.

ja-odur replied:
@d-rita Please rename the function.

d-rita replied:
noted

ja-odur commented on authors/apps/authentication/tests_/test_base_class.py:
```diff
@@ -14,6 +14,13 @@ def setUp(self):
                 "password": "JakeJake12"
             }
         }
+        self.user_data_invalid_username = {
```
@d-rita we have test data folder. this can be added there to prevent over-crowding the `base_test_case`

MuhweziDeo reviewed (changes requested):
@d-rita  I posted a username with spaces it didnt raise any issues it actually saved the user
![screen shot 2019-01-21 at 12 22 36 pm](https://user-images.githubusercontent.com/39124174/51464801-77d0d800-1d77-11e9-95c9-c1509787a33d.png)

d-rita commented:
@MuhweziDeo i have worked on what you said

d-rita committed:
fix(signup): Validate username
- Validate length of username
- Validate against username starting with a number
- Validate against special characters in username
- Write descriptive error messages
- Write tests to test validation functionality
- Resolve conflicts
[Fixes #163298000]

mariamiah commented:
@d-rita i think the username should not be accepted if it contains numbers only, please look into that

mariamiah commented:
<img width="686" alt="screen shot 2019-01-21 at 4 13 43 pm" src="https://user-images.githubusercontent.com/39744587/51476798-a363ba80-1d97-11e9-97c1-8a1d1b79e5d1.png">
Check this message about the password and change it to username

ja-odur reviewed (dismissed):
LGTM

mariamiah commented:
Looks Good To Me

kdaglas reviewed (approved):
ok, the regex now works fine, good work @d-rita

ja-odur reviewed (approved):
LGTM
